### DISCUSSION TITLE
[FIXED] Avoid MacOS 12.3

### DISCUSSION THREAD

**Discussion by milljm** [idaholab/moose#20564] [News]
UPDATE: This issue has been resolved with updates to the MOOSE conda packages. Please take a look at Discussion announcement #20630. Thanks! 

A recent Discussion revealed a possible issue when using the latest version of MacOS (12.3):

```pre
mpiexec -n 2 <your application>
<trimmed>
libpmpi.12.dylib                    0x000000010cae3994 MPIR_Assert_fail + 52
Error in system call pthread_mutex_destroy: Invalid argument src/mpi/debugger/dbginit.c:189
<trimmed>
```

Until we investigate a solution it is advised users refrain from upgrading to this version of MacOS.

**Comment by roystgnr** [comment]
Does that take a full Moose run to trigger it, or can it be replicated with just libMesh (or just TIMPI?  just some_mpi_hello_world.c) too?

There's definitely enough Undefined Behavior in [the `pthread_mutex_destroy` man page](https://pubs.opengroup.org/onlinepubs/007904875/functions/pthread_mutex_destroy.html) that it's hard to be certain this is an Xcode bug - if we (or more likely MPI, if it triggers without us even setting `--n_threads=` ?) somehow try to destroy a mutex that's still in use anywhere, it'd be entirely valid for some pthreads implementations to let us get away with it while other implementations scream and die.

**Reply by milljm** [reply]
Internal 'Conda' testing was able to trigger it during a recent `conda mambabuild .` for the moose-mpich package:

```pre
MPICH configure: 	--prefix=$PREFIX --enable-shared --enable-sharedlibs=clang --enable-fast=O2 --enable-debuginfo --enable-two-level-namespace --with-device=ch3 CC=x86_64-apple-darwin13.4.0-clang CXX=x86_64-apple-darwin13.4.0-clang++ FC=x86_64-apple-darwin13.4.0-gfortran F77=x86_64-apple-darwin13.4.0-gfortran F90= CFLAGS=-march=core2 -mtune=haswell -I$PREFIX/include CXXFLAGS=-march=core2 -mtune=haswell -I$PREFIX/include FFLAGS=-march=core2 -mtune=haswell -I$PREFIX/include LDFLAGS=-L$PREFIX/lib -Wl,-rpath,$PREFIX/lib FCFLAGS=-march=core2 -mtune=haswell -I$PREFIX/include F90FLAGS= F77FLAGS=
MPICH CC: 	x86_64-apple-darwin13.4.0-clang -march=core2 -mtune=haswell -I$PREFIX/include   -O2
MPICH CXX: 	x86_64-apple-darwin13.4.0-clang++ -march=core2 -mtune=haswell -I$PREFIX/include  -O2
MPICH F77: 	x86_64-apple-darwin13.4.0-gfortran -march=core2 -mtune=haswell -I$PREFIX/include  -O2
MPICH FC: 	x86_64-apple-darwin13.4.0-gfortran -march=core2 -mtune=haswell -I$PREFIX/include  -O2
MPICH Custom Information: 	
$PREFIX/bin/mpiexec
Error in system call pthread_mutex_destroy: Invalid argument
    src/mpi/debugger/dbginit.c:189
Assertion failed in file src/mpi/debugger/dbginit.c at line 190: err == 0
0   libpmpi.12.dylib                    0x000000010963e771 MPL_backtrace_show + 49
1   libpmpi.12.dylib                    0x000000010959e9c4 MPIR_Assert_fail + 36
2   libpmpi.12.dylib                    0x0000000109eba13b MPIR_Debugger_set_aborting + 107
3   libpmpi.12.dylib                    0x0000000109608576 MPID_Abort + 198
4   libpmpi.12.dylib                    0x0000000109597302 MPIR_Err_return_comm + 482
5   mpivars                             0x00000001084836ce main + 46
6   dyld  
```

**Reply by milljm** [reply]
FYI, I updated the Discussion. It turns out this is due to a MacOS update, not Xcode by itself.
